text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,411 @@
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Templating values of `CARGO_TARGET_DIR`
+
+Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-base-dir = "/absolute/path/to/cache/{manifest-path-hash}"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIR="/absolute/path/to/cache/{manifest-path-hash}" cargo build`
+- Through the command line flag: `cargo build --target-dir "/absolute/path/to/cache/{manifest-path-hash}"`
+
+## Naming
+
+In the example in the previous section, using a templated `CARGO_TARGET_DIR` with `cargo build` produces named subdirectories. The name of those is computed from the full and canonicalized path to the manifest for the workspace, so building `work-project/crate-1` will still use the directory for the whole workspace during a `cargo build` call.
+
+This naming scheme is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
+
+The path used for the naming of the final target directory is the one found *after* symlink resolution: `bazel` does it too and I have not found any complaints about this and it has the distinct advantage of allowing to make a symlink to a project somewhere else on the machine (for example to organise work projects) and avoid duplicating the build directory and all its data (which can be quite heavy).
+
+To prevent collisions by craftings paths, the `<manifest-path-hash>` directory will be computed from a hash of the workspace manifest's full path (and possibly other data, for example `bazel` uses its version and the current user too).
+
+## Symbolic links
+
+In the following situation
+
+```
+/Users/
+├─ poliorcetics/
+│  ├─ projects/
+│  │  ├─ actual-crate/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ symlink-to-crate/ -> actual-crate/
+```
+
+When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `{manifest-path-hash}`, or in other words: symbolic links are resolved.
+
+### Handle possibly thousands of directories in a single templated `CARGO_TARGET_DIR` path
```

Comment:
I'm not completely convinced that having thousands of directories is likely. That would mean that the host would have 1000s of Cargo projects, without ever cleaning up.

I'd also like to see how this might interact with [Cargo's ongoing GC work](https://blog.rust-lang.org/2023/12/11/cargo-cache-cleaning.html)  covered. Does that help mitigate the concerns?

Reply:
See rust-lang/cargo#13136 which is linked to within the RFC but the link is broken

Reply:
@arlosi Many packaging and build systems have the same pattern. Examples are following, but not limited to: Xcode, MacPorts, FreeBSD ports, yay, Gentoo Portages.

Reply:
> I'm not completely convinced that having thousands of directories is likely. That would mean that the host would have 1000s of Cargo projects, without ever cleaning up.

Entirely possible in CIs, and even local systems: on my previous machine, I used a common base for all my target directories, with a scheme like `CARGO_TARGET_DIR=/common/path/(basename (pwd))` and just ignored it for my saves. I ended up with more than 250 target dirs from various projects over time (most I uninstalled at some point, the project for garbage collection will be awesome), I can easily see someone reaching the thousands